# Homework 6

## Your Name Here (or your names here if you are pair programming)
Tori Garfield teg755
Veronica Alejandro vaa678

## Logistic Regression without using any libraries. 

In [2]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

# Your allowed to use only the above libraries that are imported. No other libs should be used in this assignment. 

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [3]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (4 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data, and implement logistic regression model ***without*** using any libs than imported above. You must do it manually.

* Do maximum **100 iterations**
* Use a very small learning rate for checking your GD implementation. 
* Your are allowed to use your choice of learning rate, like using 0.0001, 0.001 or 0.01 or 0.1 or higher/lower. 
* Visualize your costs.
* No need to add an y-intercept in this task. 


In [3]:
# Add your code Here! 
# Split dataset into two parts: 80% of 303 is 242, 20% is 61

xTrain = [[118, 122, 145, 149, 186]]
yTrain = [1, 2, 3, 4, 5]

def sigmoid(z):
    # input: continuous value
    # output: value in range from 0 to 1

    g = 1 / (1 + np.exp(-z))
    return g

def costFunction(X, y, weights): 
    # computes cost for logistic regression
    # input: X is a 2D array (training), y is a 1D array (target value for training),
    # weights is a 1D array of weights / fitting parameters
    # output: cost of using theta as the parameter for linear regression to fit the data points in X and y
    
    n, d = X.shape
    x_dot_weights = X.dot(weights)
    cost = 1.0 / n * (-y.T.dot(np.log(sigmoid(x_dot_weights))) - (1 - y).T.dot(np.log(1 - sigmoid(x_dot_weights))))
    return cost

def gradient(X, y, weights): 
    # computes gradient for logistic regression
    # input: same as above
    # output: grad is an np array ; gradient of the cost with respect to the parameters theta
    
    n, d = X.shape
    x_dot_weights = X.dot(weights)
    grad = (1.0 / n) * (sigmoid(x_dot_weights) - y).T.dot(X)
    return grad

weights = np.zeros(3)
numIter = 100
learnRate = 0.00001
costList = []

for i in range(0, numIter):
    
    # calculate cost: 
    cost = costFunction(xTrain, yTrain, weights) 
    
    # print and save cost
    print("#{}. Cost is: {}".format(i + 1, cost))
    costList.append(cost)
    
    # calculate the gradients
    grad = gradient(xTrain, yTrain, weights)
    
    # use the gradient to update the weights  
    weights = weights - grad * learnRate

    
plt.plot(np.arange(numIter), costList)
plt.show()

AttributeError: 'list' object has no attribute 'shape'

# Task 2 - (4 points)

Cacluate the accuracy, Precision, Recall and F1 score of your logistic regression implementaion. 
Print the results. 

You may use equations shown in lecture/slides/examples.

In [5]:
# Add your code Here! 
#calculate accuracy 
def predict(weights, X):
    p = sigmoid(X.dot(weights)) >= 0.5
    return p.astype(int)

predictions = predict(weights, xTest)
correct = np.sum(predictions == yTest)
print("Accuracy is {}.".format(correct / yTest.size))

#calculate precision
#TP/TP+FP
"""
pred = predict(weights, xTest)
pred1 = np.where(pred == 1)
TP_indexes = []
for k in pred1[0]:
        if(yTest.iloc[k][0] == 1):
            TP_Indexes.append(k)
X_test_TP = xTest.iloc[TP_indexes]
"""
true_labels = df['Target']
pred_labels = predict(weights, xTest)
TP = np.sum(np.logical_and(pred_labels == 1, true_labels == 1))
FP = np.sum(np.logical_and(pred_labels == 1, true_labels == 0))
precision = TP / (TP + FP)
print("Precision is {}.".format(precision))

#calculate recall
#TP/TP+FN
FN = np.sum(np.logical_and(pred_labels == 0, true_labels == 1))
recall = TP / (TP + FN)
print("Recall is {}.".format(recall))

#calculate F1 score
#2 * ((precision*recall)/(precision + recall))
numerator = precision * recall
denom = precision + recall
F1 = 2 * (numerator / denom)
print ("The F1 score is {}.".format(F1))

Accuracy is 0.6229508196721312.
Precision is 0.4793388429752066.
Recall is 0.4549019607843137.
The F1 score is 0.46680080482897385.


# Task 3 - (4 points)


Add y-intercept and repeat the above 2 tasks. Do you see any differences after adding the y-intercept? 


In [13]:
weights = np.zeros(4)


# Task 4 - Implement the Bold Driver   - (4 points)
In your GD implementation, add the bold driver idea to have a dynamic learning rate.

* Add a stop codition to stop the GD when the cost is not changing more than 0.001. (differences between two costs not more than 0.001, then stop)
* Can you stop earlier than 100 iterations? 




In [6]:
# Add your code Here! 
weights = np.zeros(3)
numIter = 100
learnRate = 0.0001
costList = []
precision = 0.001
oldCost = 0
for i in range(0, numIter):
    
    # calculate cost: 
    cost = costFunction(xTrain, yTrain, weights) 
    
    # print and save cost
    print("Iteration: ", i)
    print("#{}. Cost is: {}".format(i + 1, cost))
    costList.append(cost)
    
    # calculate the gradients
    grad = gradient(xTrain, yTrain, weights)
    
    # use the gradient to update the weights  
    weights = weights - grad * learnRate
    
    if(abs(cost - oldCost) <= precision):
        print("Stoped at iteration", i)
        break
    if (cost < oldCost):
        learnRate = learnRate*1.05
    elif (cost > oldCost):
        learnRate = learnRate*0.5
    oldCost = cost

Iteration:  0
#1. Cost is: 0.6931471805599451
Iteration:  1
#2. Cost is: 0.6929004299526385
Stoped at iteration 1


# Task 5 - Implement the L2 norm regularization.  - (4 points)

Modify your Cost and gradient to implement the l2 norm regularization. 
Repreat the task 1 and 2 to check if your result is changing. 

* Use y-itercept. 
* Do max 100 iterations as before and report your accuracy, Precision, Recall and F1. 
* You can stop earlier when the cost is not changing than 0.001. 


**Optional:** you might want to use the bold driver. But you can do this task without the bold driver as well. 



In [ ]:
# Add your code Here! 